src/v/cluster/partition_leaders_table.cc

```diff
@@ -90,45 +94,67 @@ void partition_leaders_table::update_partition_leader(
     update_partition_leader(ntp, model::revision_id{}, term, leader_id);
 }
 
-void partition_leaders_table::update_partition_leader(
-  const model::ntp& ntp,
+ss::future<> partition_leaders_table::update_with_node_report(
+  const node_health_report& node_report) {
+    for (const auto& topic : node_report.topics) {
+        /**
+         * Here we minimize the number of topic table and topic map lookups by
+         * doing it only once for each topic.
+         **/
+        version topic_map_version_snapshot = _topic_map_version;
+        auto [t_it, _] = _topic_leaders.try_emplace(topic.tp_ns);
+
+        const bool is_controller = topic.tp_ns == model::controller_nt;
+        bool present_in_table = _topic_table.local().contains(topic.tp_ns);
+        auto last_applied_revision
+          = _topic_table.local().last_applied_revision();
+        for (auto& p : topic.partitions) {
+            if (!p.leader_id.has_value()) {
+                continue;
+            }
+
+            const auto topic_removed = p.revision_id <= last_applied_revision
+                                       && !present_in_table;
+
+            if (topic_removed && !is_controller) {
+                vlog(
+                  clusterlog.trace,
+                  "can't update leadership of the removed topic {}",
+                  topic.tp_ns);
+                break;
+            }
+            do_update_partition_leader(
+              is_controller, t_it, p.id, p.revision_id, p.term, p.leader_id);
+            co_await ss::coroutine::maybe_yield();
```

Comment:
Note that `maybe_yielding` isn't entirely cheap ([see also](https://github.com/redpanda-data/redpanda/pull/16228)).

So doing it after every partition might be a bit expensive. I think doing it for every topic or every 100 partitions or something might be acceptable.